plugins/post-transaction-actions.py

```diff
@@ -0,0 +1,158 @@
+
+        action_tuples = []  # (action key, action_state, shell command)
+        for file_name in action_file_list:
+            for line in open(file_name).readlines():
+                line = line.strip()
+                if line and line[0] != "#":
+                    try:
+                        (action_key, action_state, action_command) = line.split(':', 2)
+                    except ValueError as e:
+                        self.logger.error(_('Bad Action Line "%s": %s') % (line, e))
+                        continue
+                    else:
+                        action_tuples.append((action_key, action_state, action_command))
+
+        return action_tuples
+
+    def _replace_vars(self, ts_item, command):
+        """Replaces variables in the command.
+        Variables: ${name}, ${arch}, ${ver}, ${rel}, ${epoch}, ${repoid}, ${state}
+        or $name, $arch, $ver, $rel, $epoch, $repoid, $state"""
+
+        action_dict = {dnf.transaction.PKG_DOWNGRADE: "downgrade",
+                       dnf.transaction.PKG_DOWNGRADED: "downgraded",
+                       dnf.transaction.PKG_INSTALL: "install",
+                       dnf.transaction.PKG_OBSOLETE: "obsolete",
+                       dnf.transaction.PKG_OBSOLETED: "obsoleted",
+                       dnf.transaction.PKG_REINSTALL: "reinstall",
+                       dnf.transaction.PKG_REINSTALLED: "reinstalled",
+                       dnf.transaction.PKG_REMOVE: "remove",
+                       dnf.transaction.PKG_UPGRADE: "upgrade",
+                       dnf.transaction.PKG_UPGRADED: "upgraded"}
+
+        try:
+            action = action_dict[ts_item.action]
+        except KeyError:
+            action = "unknown - %s" % ts_item.action
+
+        vardict = {"name": ts_item.name,
+                   "arch": ts_item.arch,
+                   "ver": ts_item.version,
+                   "rel": ts_item.release,
+                   "epoch": str(ts_item.epoch),
+                   "repoid": ts_item.from_repo,
+                   "state": action}
+
+        result = libdnf.conf.ConfigParser_substitute(command, vardict)
+        return result
+
+    def transaction(self):
+        action_tuples = self._parse_actions()
+
+        forward_ts_items = []
+        backward_ts_items = []
+        all_ts_items = []
+        for ts_item in self.base.transaction:
+            if ts_item.action in dnf.transaction.FORWARD_ACTIONS:
+                forward_ts_items.append(ts_item)
+            elif ts_item.action in dnf.transaction.BACKWARD_ACTIONS:
+                backward_ts_items.append(ts_item)
+            all_ts_items.append(ts_item)
+
+        commands_to_run = set()
+        for (a_key, a_state, a_command) in action_tuples:
+            if a_state == "forward":
+                ts_items = forward_ts_items
+            elif a_state == "backward":
+                ts_items = backward_ts_items
+            elif a_state == "any":
+                ts_items = all_ts_items
```

Comment:
You do not need all_ts_items list at all, you can use `ts_items = forward_ts_items + backward_ts_items` (in case the order of items is not important)

Reply:
Using all_ts_items is optimization. The `all_ts_items` is prepared once. Instead of making sum of lists in each loop iteration.